bin/ch-completion.bash

```diff
@@ -343,6 +442,78 @@ _ch_image_subcmd_get () {
     echo "$subcmd"
 }
 
+# Horrible, disgusting function to find an image or image ref in the ch-run
+# command line. This function takes five arguments:
+#
+#   1.) A string representing the path to the storage directory.
+#   2.) The current position of the cursor in the array representing the command
+#       line (index starting at 0).
+#   3.) A “reference” to a variable. If “_ch_run_image_finder” finds the name of
+#       an image in storage (e.g. “alpine:latest”) or something that looks like an
+#       image path (i.e. a directory, tarball or file named like a squash archive)
+#       in the command line, the value of the variable will be updated to the image
+#       name or path. If neither are found, the function will not modify the value
+#       of this variable.
+#   4.) Another “reference” to a variable. If this function finds “--” in the
+#       current command line and it doesn't seem like the user is trying to
+#       complete that “--” to an option, “_ch_run_image_finder” will assume that
+#       this is the point beyond which the user specifies commands to be run inside
+#       the container and will give the variable the index value of the “--”.
+#   5.) A string representing the expanded command line array (i.e. "${array[@]}").
+#
+# Note that when I say variable “reference” here I mean the unquoted name of the
+# variable in question (i.e. “var” instead of “$var” or “"$var"”). Passing the
+# variables to the function in this way allows it to change their values, and for
+# those changes to persist in the scope that called the function.
+#
+_ch_run_image_finder () {
+    # Takes array of words. Tries to find an image in there.
+    local images                   # these two lines are separate b/c SC2155
+    images=$(_ch_list_images "$1") #
+    shift 1
+    local cword="$1"
+    shift 1
+    local -n cli_img=$1
+    local -n cmd_pt=$2
+    shift 2
+    local wrds=("$@")
+    local ct=1
+
+    # Check for tarballs and squashfs archives.
+    while ((ct < ${#wrds[@]})); do
+        # In bash, expansion of the “~” character to the value of $HOME doesn't
+        # happen if a value is quoted (see
+        # https://stackoverflow.com/a/52519780). To work around this, we add
+        # “eval echo” (https://stackoverflow.com/a/6988394) to this test.
```

Comment:
`eval` can be very dangerous ([e.g.](https://stackoverflow.com/questions/6988288/linux-bash-how-to-unquote/6988394#comment132901076_6988394)). Probably good to work hard to avoid it.